[BUG] org.opensearch.search.simple.SimpleSearchIT.testSimpleTerminateAfterCount {p0={"search.concurrent_segment_search.enabled":"true"}}

References
https://build.ci.opensearch.org/job/gradle-check/25191/testReport/junit/org.opensearch.search.simple/SimpleSearchIT/testSimpleTerminateAfterCount__p0___search_concurrent_segment_search_enabled___true___/

To Reproduce
``` 
./gradlew ':server:internalClusterTest' --tests "org.opensearch.search.simple.SimpleSearchIT" -Dtests.method="testSimpleTerminateAfterCount {p0={"search.concurrent_segment_search.enabled":"true"}}" -Dtests.seed=BF79A24A152C3CB1 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=th -Dtests.timezone=America/Danmarkshavn -Druntime.java=20
```


Stacktrace
```
java.lang.AssertionError: Count is 1+ hits but 1 was expected.  Total shards: 1 Successful shards: 1 & 0 shard failures:
	at __randomizedtesting.SeedInfo.seed([BF79A24A152C3CB1:45F350BC58E7D135]:0)
	at org.junit.Assert.fail(Assert.java:89)
	at org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount(OpenSearchAssertions.java:303)
	at org.opensearch.search.simple.SimpleSearchIT.testSimpleTerminateAfterCount(SimpleSearchIT.java:305)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
	at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.base/java.lang.Thread.run(Thread.java:1623)

```

I am able to reproduce this using the test seed:
```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.search.simple.SimpleSearchIT" -Dtests.method="testSimpleTerminateAfterCount {p0={"search.concurrent_segment_search.enabled":"false"}}" -Dtests.seed=C37FD3C5A66D02BC -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=he-IL -Dtests.timezone=Asia/Qostanay -Druntime.java=20

org.opensearch.search.simple.SimpleSearchIT > testSimpleTerminateAfterCount {p0={"search.concurrent_segment_search.enabled":"false"}} FAILED
    java.lang.AssertionError: Count is 7 hits but 3 was expected.  Total shards: 1 Successful shards: 1 & 0 shard failures:
        at __randomizedtesting.SeedInfo.seed([C37FD3C5A66D02BC:39F52133EBA6EF38]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount(OpenSearchAssertions.java:303)
        at org.opensearch.search.simple.SimpleSearchIT.testSimpleTerminateAfterCount(SimpleSearchIT.java:311)

REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.search.simple.SimpleSearchIT" -Dtests.method="testSimpleTerminateAfterCount {p0={"search.concurrent_segment_search.enabled":"true"}}" -Dtests.seed=C37FD3C5A66D02BC -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=he-IL -Dtests.timezone=Asia/Qostanay -Druntime.java=20

org.opensearch.search.simple.SimpleSearchIT > testSimpleTerminateAfterCount {p0={"search.concurrent_segment_search.enabled":"true"}} FAILED
    java.lang.AssertionError: Count is 4 hits but 3 was expected.  Total shards: 1 Successful shards: 1 & 0 shard failures:
        at __randomizedtesting.SeedInfo.seed([C37FD3C5A66D02BC:39F52133EBA6EF38]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount(OpenSearchAssertions.java:303)
        at org.opensearch.search.simple.SimpleSearchIT.testSimpleTerminateAfterCount(SimpleSearchIT.java:311)


Suite: Test class org.opensearch.search.simple.SimpleSearchIT
```

However, it looks like there is flakiness for both concurrent and non-concurrent cases (which makes sense since concurrent search is disabled for this path).

Just to be sure that this is an existing problem from before concurrent search changes, I checked out the 2.7 branch and applied the `setSize(0)` and saw the same failure:
```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.search.simple.SimpleSearchIT.testSimpleTerminateAfterCount" -Dtests.seed=C37FD3C5A66D02BC -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=he-IL -Dtests.timezone=Asia/Qostanay -Druntime.java=17

org.opensearch.search.simple.SimpleSearchIT > testSimpleTerminateAfterCount FAILED
    java.lang.AssertionError: Count is 8 hits but 3 was expected.  Total shards: 1 Successful shards: 1 & 0 shard failures:
        at __randomizedtesting.SeedInfo.seed([C37FD3C5A66D02BC:39F52133EBA6EF38]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount(OpenSearchAssertions.java:306)
        at org.opensearch.search.simple.SimpleSearchIT.testSimpleTerminateAfterCount(SimpleSearchIT.java:284)
```

--- a/server/src/internalClusterTest/java/org/opensearch/search/simple/SimpleSearchIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/search/simple/SimpleSearchIT.java
@@ -283,10 +283,9 @@ public class SimpleSearchIT extends ParameterizedOpenSearchIntegTestCase {
         assertHitCount(searchResponse, 2L);
     }
 
-    public void testSimpleTerminateAfterCount() throws Exception {
+    public void dotestSimpleTerminateAfterCountWithSize(int size, int max) throws Exception {
         prepareCreate("test").setSettings(Settings.builder().put(SETTING_NUMBER_OF_SHARDS, 1).put(SETTING_NUMBER_OF_REPLICAS, 0)).get();
         ensureGreen();
-        int max = randomIntBetween(3, 29);
         List<IndexRequestBuilder> docbuilders = new ArrayList<>(max);
 
         for (int i = 1; i <= max; i++) {
@@ -299,9 +298,7 @@ public class SimpleSearchIT extends ParameterizedOpenSearchIntegTestCase {
         refresh();
 
         SearchResponse searchResponse;
-        int size;
         for (int i = 1; i < max; i++) {
-            size = randomIntBetween(0, max);
             searchResponse = client().prepareSearch("test")
                 .setQuery(QueryBuilders.rangeQuery("field").gte(1).lte(max))
                 .setTerminateAfter(i)
@@ -322,7 +319,18 @@ public class SimpleSearchIT extends ParameterizedOpenSearchIntegTestCase {
         assertFalse(searchResponse.isTerminatedEarly());
     }
 
-    public void testSimpleTerminateAfterTrackTotalHitsUpTo() throws Exception {
+    @AwaitsFix(bugUrl = "https://github.com/opensearch-project/OpenSearch/issues/10435")
+    public void testSimpleTerminateAfterCountSize0() throws Exception {
+        int max = randomIntBetween(3, 29);
+        dotestSimpleTerminateAfterCountWithSize(0, max);
+    }
+
+    public void testSimpleTerminateAfterCountRandomSize() throws Exception {
+        int max = randomIntBetween(3, 29);
+        dotestSimpleTerminateAfterCountWithSize(randomIntBetween(1, max), max);
+    }
+
+    public void doTestSimpleTerminateAfterTrackTotalHitsUpTo(int size) throws Exception {
         prepareCreate("test").setSettings(Settings.builder().put(SETTING_NUMBER_OF_SHARDS, 1).put(SETTING_NUMBER_OF_REPLICAS, 0)).get();
         ensureGreen();
         int numDocs = 29;
@@ -337,8 +345,6 @@ public class SimpleSearchIT extends ParameterizedOpenSearchIntegTestCase {
         ensureGreen();
         refresh();
 
-        // size=0 is a special case where topDocsCollector is not added
-        int size = randomIntBetween(0, 1);
         SearchResponse searchResponse;
         searchResponse = client().prepareSearch("test")
             .setQuery(QueryBuilders.rangeQuery("field").gte(1).lte(numDocs))
@@ -397,6 +403,15 @@ public class SimpleSearchIT extends ParameterizedOpenSearchIntegTestCase {
             .get();
         assertEquals(5, searchResponse.getHits().getTotalHits().value);
         assertEquals(GREATER_THAN_OR_EQUAL_TO, searchResponse.getHits().getTotalHits().relation);
+    }
+
+    @AwaitsFix(bugUrl = "https://github.com/opensearch-project/OpenSearch/issues/10435")
+    public void testSimpleTerminateAfterTrackTotalHitsUpToRandomSize() throws Exception {
+        doTestSimpleTerminateAfterTrackTotalHitsUpTo(0);
+    }
+
+    public void testSimpleTerminateAfterTrackTotalHitsUpToSize0() throws Exception {
+        doTestSimpleTerminateAfterTrackTotalHitsUpTo(randomIntBetween(1, 29));
     }
 
     public void testSimpleIndexSortEarlyTerminate() throws Exception {

